x/oauth2cors/cors.go

```diff
@@ -21,113 +19,129 @@ import (
-		if o == "*" {
-			alwaysAllow = true
-		}
-		// if the protocol (http or https) is specified, but the url is wildcard, use special ** glob, which ignore the '.' separator.
-		// This way g := glob.Compile("http://**") g.Match("http://google.com") returns true.
-		if splittedO := strings.Split(o, "://"); len(splittedO) != 1 && splittedO[1] == "*" {
-			o = fmt.Sprintf("%s://**", splittedO[0])
-		}
-		g, err := glob.Compile(strings.ToLower(o), '.')
-		if err != nil {
-			reg.Logger().WithError(err).Fatalf("Unable to parse cors origin: %s", o)
-		}
-
-		patterns = append(patterns, g)
-	}
-
-	options := cors.Options{
-		AllowedOrigins:     opts.AllowedOrigins,
-		AllowedMethods:     opts.AllowedMethods,
-		AllowedHeaders:     opts.AllowedHeaders,
-		ExposedHeaders:     opts.ExposedHeaders,
-		MaxAge:             opts.MaxAge,
-		AllowCredentials:   opts.AllowCredentials,
-		OptionsPassthrough: opts.OptionsPassthrough,
-		Debug:              opts.Debug,
-		AllowOriginRequestFunc: func(r *http.Request, origin string) bool {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 			ctx := r.Context()
-			if alwaysAllow {
-				return true
-			}
 
-			origin = strings.ToLower(origin)
-			for _, p := range patterns {
-				if p.Match(origin) {
-					return true
-				}
+			opts, enabled := reg.Config().CORS(ctx, config.PublicInterface)
+			if !enabled {
+				reg.Logger().Debug("not enhancing CORS per client, as CORS is disabled")
+				h.ServeHTTP(w, r)
+				return
 			}
 
-			// pre-flight requests do not contain credentials (cookies, HTTP authorization)
-			// so we return true in all cases here.
-			if r.Method == http.MethodOptions {
-				return true
-			}
-
-			var clientID string
-
-			// if the client uses client_secret_post auth it will provide its client ID in form data
-			clientID = r.PostFormValue("client_id")
-
-			// if the client uses client_secret_basic auth the client ID will be the username component
-			if clientID == "" {
-				clientID, _, _ = r.BasicAuth()
-			}
-
-			// otherwise, this may be a bearer auth request, in which case we can introspect the token
-			if clientID == "" {
-				token := fosite.AccessTokenFromRequest(r)
-				if token == "" {
-					return false
+			alwaysAllow := len(opts.AllowedOrigins) == 0
+			patterns := make([]glob.Glob, 0, len(opts.AllowedOrigins))
+			for _, o := range opts.AllowedOrigins {
+				if o == "*" {
+					alwaysAllow = true
+					break
 				}
-
-				session := oauth2.NewSessionWithCustomClaims("", reg.Config().AllowedTopLevelClaims(ctx))
-				_, ar, err := reg.OAuth2Provider().IntrospectToken(ctx, token, fosite.AccessToken, session)
+				// if the protocol (http or https) is specified, but the url is wildcard, use special ** glob, which ignore the '.' separator.
+				// This way g := glob.Compile("http://**") g.Match("http://google.com") returns true.
+				if scheme, rest, found := strings.Cut(o, "://"); found && rest == "*" {
```

Comment:
What's the point of changing this if it worked before? Are there tests to prove that this change works?

Reply:
I assumed that this did not work as expected, I can add tests. This is just a copy from above, including the comment.

Reply:
Added tests. The client origins now behave the same as the global.s